PYIC-3591: Set up catalogue of core-front pages

AmritSidhu opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Set up catalogue of core-front pages

Use the NODE_ENV to determine whether we are in a non-production environment to display the new route
Repurposed the existing unused data we already had defined
Moved the reuse code into the helper as it is now used in two places
<img width="400" alt="Screenshot 2023-09-28 at 09 49 19" src="https://github.com/alphagov/di-ipv-core-front/assets/24409958/cc05966a-715a-42d3-a0a4-8cd48d0bc7e5">

### Why did it change

We’d like devs/testers/UCD/any non-technical stakeholders to be able to view a catalogue of all the pages served by core-front, labelled by their path slug/file name with each page clickable to view. This is also important so we develop a common language around page names with UCD

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-3591](https://govukverify.atlassian.net/browse/PYIC-3591)

## Checklists

### Environment variables or secrets

<!-- Delete if changes DO include new environment variables or secrets -->
- [x] No environment variables or secrets were added or changes


[PYIC-3591]: https://govukverify.atlassian.net/browse/PYIC-3591?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

AmritSidhu committed:
PYIC-3591: Set up catalogue of core-front pages

Use the NODE_ENV to determine whether we are in a non-production environment to display the new route

AmritSidhu committed:
PYIC-3591: Use local config specifically for the reuse page

Repurposed the existing unused data we already had defined
Refactored to move the reuse code into the helper as it is now used in two places

AmritSidhu commented on src/app/ipv/router.js:
```diff
@@ -27,6 +28,11 @@
 
+const currentEnvironment = process.env.NODE_ENV;
+if (currentEnvironment === "development") {
+  router.get("/all-templates", allTemplates);
```
Dismissed as this route is only accessible in dev/build

AmritSidhu committed:
PYIC-3591: Add unit tests for reuseHelper

Tests check that the data is in an expected structure

AmritSidhu committed:
PYIC-3591: Add a visistedAllTemplate flag

If a user has visisted the all-templates url then they will be able to visit any page they want

RogVanGDS commented on src/views/ipv/all-templates.njk:
```diff
@@ -0,0 +1,24 @@
+    <thead class="govuk-table__head">
+      <tr class="govuk-table__row">
+        <th scope="col" class="govuk-table__header">List of all pages in di-ipv-core-front</th>
+        <th scope="col" class="govuk-table__header"></th>
```
Is it worth having this as a larger header than the content rows below? (Know that this isn't user facing, but might be better for readability nonetheless

RogVanGDS commented on src/app/ipv/middleware.js:
```diff
@@ -187,7 +192,31 @@ module.exports = {
       const { pageId } = req.params;
+
+      if (
+        currentEnvironment === "development" &&
```
Might be worth having "development" as a constant or an Enum value to avoid accidental changes in the future?

RogVanGDS commented on src/app/ipv/middleware.js:
```diff
@@ -187,7 +192,31 @@ module.exports = {
+        req.session.visitedAllTemplates
+      ) {
+        if (pageId === "page-ipv-reuse") {
+          let userDetailsResponse = samplePersistedUserDetails;
```
Does this need to be `let`? (I may have missed the reassignment in all fairness)

RogVanGDS commented on src/app/ipv/middleware.js:
```diff
@@ -187,7 +192,31 @@ module.exports = {
+        if (pageId === "page-ipv-reuse") {
+          let userDetailsResponse = samplePersistedUserDetails;
+          const i18n = req.i18n;
+          const userDetails = generateUserDetails(userDetailsResponse, i18n);
```
Do we lose much readability by just having `const userDetails = generateUserDetails(samplePersistedUserDetails, req.i18n);` here and removing the above two lines?

RogVanGDS reviewed (dismissed):
Looks good to me, feel free to ignore me being a nit-pick if you want 🙂

AmritSidhu committed:
PYIC-3591: Refactor and tidy up

Put the env name under a constant variable
Remove unecessary assignments
Use const over let
Make html title larger

RogVanGDS reviewed (approved):
LGTM
